feat: integrate got-ocr2.0 as image reader

phv2312 opened the pull request:
## Description

- Integrate the got-ocr2.0 OCR as image reader
- New extension manager for easily switch between different supported loaders
- Also, thanks @cin-jimmy for his suggestion on github stale (issue)

## Type of change

- [x] New features (non-breaking change).
- [ ] Bug fix (non-breaking change).
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected).

## Checklist

- [x] I have performed a self-review of my code.
- [ ] I have added thorough tests if it is a core feature.
- [ ] There is a reference to the original bug report and related work.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] The feature is well documented.

phv2312 committed:
feat: update stale action

phv2312 committed:
feat: comfort gocr2 format

phv2312 committed:
feat: update state for extension manager while load setting

phv2312 committed:
feat: comfort pre-commit

phv2312 committed:
feat: update ocr loader

phv2312 committed:
feat: change default loader for image

phv2312 committed:
feat: update guideline

phv2312 committed:
feat: update github stales & remove unncessary files

cin-niko commented:
@phv2312, can you add a docker-compose file (allow choose the docker image for OCR service)? I think it will help people test more easily.

taprosoft commented on libs/kotaemon/kotaemon/indices/ingests/extensions.py:
```diff
@@ -0,0 +1,140 @@
+)
+
+unstructured = UnstructuredReader()
+adobe_reader = AdobeReader()
```
@phv2312 let add AdobeReader and AzureDIReader as selectable options on Extensions UI. Also remove it from Retrieval settings since it is no longer needed :D

ngduyanhece commented on libs/kotaemon/kotaemon/indices/ingests/extensions.py:
```diff
@@ -0,0 +1,140 @@
+from kotaemon.loaders import (
+    AdobeReader,
+    AzureAIDocumentIntelligenceLoader,
+    GOCR2ImageReader,
```
**Suggestion:** Add error handling during the initialization of `GOCR2ImageReader`.

**Existing Code:**
```Python
GOCR2ImageReader,
```

**Improved Code:**
```Python
try:
    self.gocr2_reader = GOCR2ImageReader()
except Exception as e:
    raise RuntimeError(f"Failed to initialize GOCR2ImageReader: {e}")
```

**Details:**
Ensure that the `GOCR2ImageReader` is properly initialized with error handling to manage potential issues during instantiation.

phv2312 replied:
In the current implementation and your suggestion. Both raise error when initialization failed. What's the difference here ?

ngduyanhece commented on integration/got-ocr2.md:
````diff
@@ -0,0 +1,30 @@
+- We provide the docker image, with fastapi for serving the GOT-OCR2.0. Pull the image from:
+
+```bash
+docker run -d --gpus all -p 8881:8881 ghcr.io/phv2312/got-ocr2.0:main
````
**Suggestion:** Clarify setup instructions for running the Docker image.

**Existing Code:**
```Markdown
docker run -d --gpus all -p 8881:8881 ghcr.io/phv2312/got-ocr2.0:main
```

**Improved Code:**
```Markdown
# Ensure Docker is installed and running, and that you have GPU support if required.
# docker run -d --gpus all -p 8881:8881 ghcr.io/phv2312/got-ocr2.0:main
```

**Details:**
Clarify the setup instructions by specifying any prerequisites for running the Docker image.

phv2312 replied:
Thanks for your suggestion but It seems you're using ChatGPT for suggestion. This one make no sense at all

phv2312 committed:
feat: resolve conflicts

phv2312 committed:
feat: update extentions for pdf

phv2312 committed:
feat: introduce docker-compose

phv2312 commented:
Hi @taprosoft @cin-niko. Sorry for no update for long time. Can you help to review this PR again

niko-nnkn commented on docker-compose.dev.yml:
why do we need two docker-compose files?

phv2312 replied:
One for dev and one for production @cin-niko

niko-nnkn commented on docker-compose.yml:
same as above.

phv2312 replied:
Let add docs in README to guide user how to develop with docker

niko-nnkn commented on integration/got-ocr2.md:
````diff
@@ -0,0 +1,30 @@
+- We provide the docker image, with fastapi for serving the GOT-OCR2.0. Pull the image from:
+
+```bash
+docker run -d --gpus all -p 8881:8881 ghcr.io/phv2312/got-ocr2.0:main
````
Should we move the docker to Kotaemon ghcr if we use GOCR as the main OCR loader?

niko-nnkn commented on libs/kotaemon/kotaemon/indices/ingests/extensions.py:
```diff
@@ -0,0 +1,142 @@
+    UnstructuredReader,
+)
+
+unstructured = UnstructuredReader()
```
for consistent naming (same as in `files.py`): `unstructured_reader`

niko-nnkn commented on libs/kotaemon/kotaemon/indices/ingests/extensions.py:
```diff
@@ -0,0 +1,142 @@
+    credential=str(config("AZURE_DI_CREDENTIAL", default="")),
+    cache_dir=getattr(flowsettings, "KH_MARKDOWN_OUTPUT_DIR", None),
+)
+adobe_reader.vlm_endpoint = azure_reader.vlm_endpoint = getattr(
```
why don't we pass `vlm_endpoint` in `__init__` (same as in `files.py`)

niko-nnkn commented on libs/kotaemon/kotaemon/indices/ingests/extensions.py:
```diff
@@ -0,0 +1,142 @@
+
+        raise Exception(f"can not find the selected loader for extension: {extension}")
+
+    def generate_gradio_settings(self) -> dict[str, Any]:
```
IMO, the logic related to gradio UI should be implemented in ktem.

phv2312 committed:
fix: resolve conflicts

phv2312 committed:
fix: resolve conflicts

phv2312 committed:
chore: refactor by pre-commit

phv2312 committed:
feat: update docling reader into extension manager

phv2312 committed:
feat: bring extension manager to kotaemon

phv2312 commented:
Hi @cin-niko and @taprosoft . I have updated according to niko's comments and rebased from the latest master already.
Can you help to check this PR again ?

cin-niko committed:
refactor: move exteions manager to ktem

cin-niko commented:
@phv2312 Overall is good. But it seems that setting the loader for extensions feature doesn't work. 
For example:
- Set pdf loader in `Settings -> Retrieval Settings -> File loader`: Work
- Set pdf loader in `Settings -> Loader settings -> Loader .pdf`: Doesn't work

taprosoft commented:
@phv2312 sorry for the late comment. Overall the logic is fine but the current settings UI is a bit cluttered. I will push a small change to improve this prior to merging.

cin-niko committed:
fix: remove duplicate loader setting in retrieval settings

heapsoftware commented:
Need OCR for PDFs working as well, when I upload a PDF which is created by a scanner that scans in pages, it puts the scanned pages into a multi-page PDF. Need to be able to upload the PDF, which then it does an OCR to extract the text and use that extracted text for reasoning.
